### PR TITLE
Driver/tps2012

### PR DESCRIPTION
**Pull request by WilliamHPNielsen**
Feature: Add driver for TPS2012

Changes proposed in this pull request:
- Add a driver for the TPS2012 oscilloscope
- Add a notebook with an example of usage

Future improvements:
- Implement more features (in particular: oscilloscope measurements)

@giulioungaretti

### CODE REVIEW

File: qcodes/instrument_drivers/tektronix/TPS2012.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+                Default: False.
+        """
+        original_timeout = self.visa_handle.timeout
+        self.visa_handle.timeout = 1000  # 1 second as VISA counts in ms
+        gotexception = False
+        while not gotexception:
+            try:
+                message = self.visa_handle.read()
+                if verbose:
+                    print(message)
```
Comment (giulioungaretti) [inline qcodes/instrument_drivers/tektronix/TPS2012.py]: Maybe logging here ?

Comment (WilliamHPNielsen) [inline qcodes/instrument_drivers/tektronix/TPS2012.py]: No, I think printing is better. We talked about this already when I put the exact same function in the AWG driver (but actually, I think this function should be promoted to be in the VisaInstrument class), but in this case, verbosity is really for the user to read in his/her terminal what was still in the queue.

File: qcodes/instrument_drivers/tektronix/TPS2012.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+                           vals=vals.Enum(5e-9, 10e-9, 25e-9, 50e-9, 100e-9,
+                                          250e-9, 500e-9, 1e-6, 2.5e-6, 5e-6,
+                                          10e-6, 25e-6, 50e-6, 100e-6, 250e-6,
+                                          500e-6, 1e-3, 2.5e-3, 5e-3, 10e-3,
+                                          25e-3, 50e-3, 100e-3, 250e-3, 500e-3,
+                                          1, 2.5, 5, 10, 25, 50))
+
+        # channel-specific parameters
+
+        def curveasker(ch):
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/tektronix/TPS2012.py]: Should this be private?

File: qcodes/instrument_drivers/tektronix/TPS2012.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    ##################################################
+
+    def _binaryparser(self, curve):
+        """
+        Helper function for parsing the curve data
+
+        Args:
+            curve (str): the return value of 'CURVe?' when
+            DATa:ENCdg is set to RPBinary.
+            Note: The header and findal newline character
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/tektronix/TPS2012.py]: Typo?

File: qcodes/instrument_drivers/tektronix/TPS2012.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        outdict['no_of_bytes'] = response[0]
+        outdict['no_of_bits'] = response[1]
+        outdict['encoding'] = response[2]
+        outdict['binary_format'] = response[3]
+        outdict['byte_order'] = response[4]
+        outdict['no_of_points'] = response[5]
+        outdict['waveform_ID'] = response[6]
+        outdict['point_format'] = response[7]
+        outdict['x_incr'] = response[8]
+        # outdict['point_offset'] = response[9]  # Always zero
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/tektronix/TPS2012.py]: remove?

File: qcodes/instrument_drivers/tektronix/TPS2012.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        xdata = np.linspace(xstart, len(ydata)*xinc+xstart, len(ydata))
+        return (xdata, ydata)
+
+    ##################################################
+    # METHODS FOR THE USER                           #
+    ##################################################
+
+    def clear_message_queue(self, verbose=False):
+        """
+        Function to clear up (flush) the VISA message queue of the AWG
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/tektronix/TPS2012.py]: AWG -> TPS

File: qcodes/instrument_drivers/tektronix/TPS2012.py (hunk 4 above)
Comment (WilliamHPNielsen) [inline qcodes/instrument_drivers/tektronix/TPS2012.py]: I intentionally left it there as an explanation to curious souls; what about field 9? If you prefer, I can reformulate the line.

File: qcodes/instrument_drivers/tektronix/TPS2012.py (hunk 5 above)
Comment (WilliamHPNielsen) [inline qcodes/instrument_drivers/tektronix/TPS2012.py]: Haha, that's right! The AWG has messed my mind.

File: qcodes/instrument_drivers/tektronix/TPS2012.py (hunk 4 above)
Comment (jenshnielsen) [inline qcodes/instrument_drivers/tektronix/TPS2012.py]: It's fine I just have an allergy to commented out lines of code

Comment (giulioungaretti) [inline qcodes/instrument_drivers/tektronix/TPS2012.py]: @jenshnielsen @WilliamHPNielsen  make it a comment, i.e. start with the actual "human language and then the lien of code :D

**Comment by jenshnielsen** [comment]
I think the following should work but fails because the returned data it's setup correctly as an array type parameter. 

```python
me1 = qc.measure.Measure(tps.ch1_curvedata)
me1.run()
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-24-e70bddbfcdcf> in <module>()
----> 1 me1.run()

/Users/jhn/src/Qcodes/qcodes/measure.py in run(self, use_threads, quiet, data_manager, station, **kwargs)
     85         # run the measurement as if it were a Loop
     86         self._dummyLoop.run(background=background, use_threads=use_threads,
---> 87                             station=station, quiet=True)
     88 
     89         # look for arrays that are unnecessarily nested, and un-nest them

/Users/jhn/src/Qcodes/qcodes/loops.py in run(self, background, use_threads, quiet, data_manager, station, progress_interval, *args, **kwargs)
    853                     del self.process
    854 
--> 855                 self._run_wrapper()
    856 
    857                 if self.data_set.mode != DataMode.LOCAL:

/Users/jhn/src/Qcodes/qcodes/loops.py in _run_wrapper(self, *args, **kwargs)
    906     def _run_wrapper(self, *args, **kwargs):
    907         try:
--> 908             self._run_loop(*args, **kwargs)
    909         except _QuietInterrupt:
    910             pass

/Users/jhn/src/Qcodes/qcodes/loops.py in _run_loop(self, first_delay, action_indices, loop_indices, current_values, **ignore_kwargs)
    977                     f(first_delay=delay,
    978                       loop_indices=new_indices,
--> 979                       current_values=new_values)
    980 
    981                     # after the first action, no delay is inherited

/Users/jhn/src/Qcodes/qcodes/actions.py in __call__(self, loop_indices, **ignore_kwargs)
    144                 out_dict[param_id] = param_out
    145 
--> 146         self.store(loop_indices, out_dict)
    147 
    148 

/Users/jhn/src/Qcodes/qcodes/data/data_set.py in store(self, loop_indices, ids_values)
    611             # on the server (if you hit the if statement above) or else here
    612             for array_id, value in ids_values.items():
--> 613                 self.arrays[array_id][loop_indices] = value
    614             self.last_store = time.time()
    615             if (self.write_period is not None and

/Users/jhn/src/Qcodes/qcodes/data/data_array.py in __setitem__(self, loop_indices, value)
    339         self._update_modified_range(min_li, max_li)
    340 
--> 341         self.ndarray.__setitem__(loop_indices, value)
    342 
    343     def __getitem__(self, loop_indices):

ValueError: setting an array element with a sequence.
```

**Comment by jenshnielsen** [comment]
That is probably a good test for the arrayparameter in #403

**Comment by jenshnielsen** [comment]
@WilliamHPNielsen I have a slightly tweaked version on https://github.com/jenshnielsen/qcodes/tree/driver/TPS2012 which uses the new ArrayParameter (not complete)

**Comment by WilliamHPNielsen** [comment]
This has been replaced and merged.

**Comment by jenshnielsen** [comment]
For the record this was merged via #603

**Pull request closed without merging**